Port `norm` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62711 Port `norm` kernel to structured kernels.**
* #62024 Port `prod` kernel to structured kernels.
* #61643 Port `mean` kernel to structured kernels.

Tracking issue: #55070

Differential Revision: [D30109866](https://our.internmc.facebook.com/intern/diff/D30109866)

ysiraichi committed:
Port `norm` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30109866).

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1353,8 +1372,8 @@ inline const Tensor & _any(const Tensor & result, TensorIterator & iter) {
   Tensor result;
 
-  auto out_dtype =
-      meta::check_allany_and_get_output_dtype("any", self, result, false);
```
nice, i like the loss of the boolean

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1198,91 +1234,74 @@ Tensor& special_logsumexp_out(const Tensor& self, IntArrayRef dims, bool keepdim
+    bool keepdim) {
+  // Sparse tensors need a different implementation because their values
+  // are accessed with a different API than strided tensors
+  return at::native_norm(self, p, dim, keepdim, c10::nullopt);
```
This comment feels out of date now

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -84,45 +83,40 @@ IntArrayRef optional_to_arrayref(const c10::optional<int64_t>& opt) {
+    return result.scalar_type();
+  } else {
+    return (self.scalar_type() == kByte) ? kByte : kBool;
+  }
```
It's probably worth referencing `// Refer [all, any : uint8 compatibility]` here, since this motivates the Byte specialc ase

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1198,91 +1234,74 @@ Tensor& special_logsumexp_out(const Tensor& self, IntArrayRef dims, bool keepdim
-// leave it so we support sparse tensors
 Tensor norm(const Tensor& self, const Scalar& p) {
-  return at::native::_norm(self, p);
+  return at::norm(self, p, IntArrayRef{}, false);
```
Another one for https://github.com/pytorch/pytorch/issues/50953

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -217,6 +211,48 @@ TORCH_META_FUNC2(mean, dim)
+
+  TORCH_CHECK(
+      self.layout() == Layout::Strided,
+      "norm only supports strided layout, but got: ", self.layout());
```
Technically, now that the functions calling these are no longer composite, this test isn't needed anymore, since sparse will route to SparseCPU and there won't be any kernel.

ysiraichi committed:
Update on "Port `norm` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D30109866](https://our.internmc.facebook.com/intern/diff/D30109866)

[ghstack-poisoned]

ezyang committed:
Update on "Port `norm` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D30109866](https://our.internmc.facebook.com/intern/diff/D30109866)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30109866).
